keps/sig-scheduling/20190312-pod-affinity-and-anti-affinity-support-gt-and-lt-operators.md

```diff
@@ -0,0 +1,171 @@
+label selectors: `In`, `NotIn`, `Exists`, `DoesNotExist`.
+
+ This is not an ideal situation if users want to put pods based on the label range. 
+ `Pod` affinity/anti-affinity support for `Gt` and `Lt` operators will give users more 
+ control.
+
+ ### Goals
+
+- `Pod` affinity/anti-affinity support for `Gt` and `Lt` operators.
+- `Gt` and `Lt` will have the same status and influence as the original operators (`In`, 
+`NotIn`, `Exists`, `DoesNotExist`).
+- `Gt` and `Lt` will work with `requiredDuringSchedulingIgnoredDuringExecution`(predicate, hard requirements) and `preferredDuringScheduling`(priority, soft requirements).
+
+ ### Non-Goals
+
+- Just add `Gt` and `Lt` operator functions without changing their original definition.
+- Not changing the behavior of other label selectors, such as `ReplicaSets`, `Daemonsets`, etc.
+
+ ## Proposal
+
+ ### User Stories
+
+ As an application developer, I want my application pods to be scheduled onto
+one node that has pod with the "foo" tag and the tag value between "20" and "30".
+
+- if we use the original `In` operator to implement, then we will write affinity 
+like this:
+ ```yaml
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: In
+            values:
+            - 20
+            - 21
+            - 22
+            - 23
+            ...
+            - 28
+            - 29
+            - 30
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+```
+
+ This is not an ideal solution. A promising solution is to provide users with `Gt` and `Lt` 
+ operators, giving users the ability to specify the scope of the tag. Users can achieve 
+ this in this way:
+ ```yaml
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: Gt
+            values:
+            - 20
+          - key: foo
+            operator: Lt
+            values:
+            - 30
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+```
+
+ ### Risks and Mitigations
+
+ Along with this feature, the biggest risk should be performance. This may also be the reason 
+ why the `Gt` and `Lt` operators are not supported when Pod affinity/anti-affinity is 
+ first proposed. In order to understand the impact of these changes, we need to understand the their performance implications.So we will perform performance/benchmark tests on this change and backward compatibility test.
+
+ As part of the mitigations, the good news is that the performance of the entire scheduler has been greatly improved (https://github.com/kubernetes/kubernetes/pull/74041#issuecomment-466191359), and the processing of Pod affinity/anti-affinity has been surprisingly Optimized(https://github.com/kubernetes/kubernetes/pull/67788).
```

Comment:
Instead of "As part the mitigation..." you may want to say "Compared to the time that inter-pod affinity was introduced, performance of the scheduler has been greatly improved..."

Reply:
Yeah, thanks.